e2e/test/FileUploadE2ETests.cs

```diff
@@ -54,6 +59,97 @@ public async Task FileUpload_X509_SmallFile_Http()
             await UploadFile(Client.TransportType.Http1, smallFile, true).ConfigureAwait(false);
         }
 
+        [TestMethod]
+        [TestCategory("LongRunning")]
+        public async Task FileUpload_SmallFile_Http_GranularSteps()
+        {
+            string filename = await GetTestFileNameAsync(FileSizeSmall).ConfigureAwait(false);
+            using FileStream fileStreamSource = new FileStream(filename, FileMode.Open, FileAccess.Read);
+            Http1TransportSettings fileUploadTransportSettings = new Http1TransportSettings();
+
+            await UploadFileManual(fileStreamSource, filename, fileUploadTransportSettings).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        [TestCategory("LongRunning")]
+        public async Task FileUpload_SmallFile_Http_GranularSteps_x509()
+        {
+            string filename = await GetTestFileNameAsync(FileSizeSmall).ConfigureAwait(false);
+            using FileStream fileStreamSource = new FileStream(filename, FileMode.Open, FileAccess.Read);
+            Http1TransportSettings fileUploadTransportSettings = new Http1TransportSettings();
+
+            await UploadFileManual(fileStreamSource, filename, fileUploadTransportSettings, true).ConfigureAwait(false);
+        }
+
+        [TestMethod]
+        [TestCategory("LongRunning")]
+        public async Task FileUpload_SmallFile_Http_GranularSteps_Proxy()
+        {
+            string filename = await GetTestFileNameAsync(FileSizeSmall).ConfigureAwait(false);
+            using FileStream fileStreamSource = new FileStream(filename, FileMode.Open, FileAccess.Read);
+            Http1TransportSettings fileUploadTransportSettings = new Http1TransportSettings()
+            {
+                Proxy = new WebProxy(Configuration.IoTHub.ProxyServerAddress)
+            };
+
+            await UploadFileManual(fileStreamSource, filename, fileUploadTransportSettings).ConfigureAwait(false);
+        }
+
+        private async Task UploadFileManual(Stream source, string filename, Http1TransportSettings fileUploadTransportSettings, bool x509auth = false)
+        {
+            await FileNotificationTestListener.InitAsync().ConfigureAwait(false);
```

Comment:
This is just a suggestion - I think the tests will read better if you separate the auth from the upload. The auth part seems to belong more in the test as you are testing upload with different auth.

Reply:
Like only add tests for 
1) SAS based granular
2) SAS based granular with Prody

?

Reply:
I wanted to have an x509 test case here simply because the transportsettings of the file upload client are a bit different depending on if you use SAS or x509, so it seemed worth testing

Reply:
No I meant add the auth logic inside the test and the common method should just have the common logic. The Upload helper method seems to do a lot.